### ISSUE TITLE
Unprioritized vehicles always cut through prioritized flow in intersection

### ISSUE THREAD

**Issue by marecabo** [eclipse/sumo#8124]
Vehicles (blue) ...
- coming from an unprioritized (`g`) flow and 
- waiting within the intersection area (between stop line and internal junction)

always force their way through a prioritized flow (cyan)
- as soon as the signal of the unprioritized flow turns red (`y` -> `r`)
- and are thus blocking the prioritized flow which still has green (`G`).

I would expect, that all unprioritized vehicles already on the intersection should wait till the prioritized flow has passed (or a sufficient gap emerges, which I deliberately prevented in my test below), despite their traffic light turning red (which is behind them and which they can't "see" anyway). 

https://user-images.githubusercontent.com/23156476/105517371-40cf8d00-5cd7-11eb-8633-4e9be6ffb06c.mp4

Scenario files: [issue.zip](https://github.com/eclipse/sumo/files/5857199/issue.zip)

**First Investigation**

I am on SUMO version 1.8.0 on Ubuntu 18.04 and tried setting the following seemingly related options, however without difference in the behavior:

- Set `jmTimegapMinor="300"`, `impatience="off"`, `jmIgnoreKeepClearTime="0"` for the vehicle type of the unprioritized flow.
- Set `keepClear="false"` for the connection which the unprioritized flow is taking.

Is this behavior intended or am I (hopefully) missing another option which could control this behavior?

Thank you for your work on SUMO!

**Comment by namdre** [comment]
The current behavior is at least somewhat intended (though with another situation in mind). Assume the following case:
- traffic turning from west to north has not managed to clear the intersection
- the next green phase is north/south (with insufficient all-red time to clear the intersection)
- in this case, it is imperative that the north/south traffic permits the stuck vehicles to leave or they will block the flow

The situation in your scenario is a novel one that was not considered in the logic so far
Most traffic signal layouts I have come across fall in 2 categories:
a) they do not permit left-turns together with an oncoming flow and vehicles cannot get stuck this way
b) after the common green phase for left turn and oncoming flow, they follow up with a clearance phase or at least an all-red phase to allow vehicles on the junction to exit.

So out of curiosity: are you modeling a real world phase layout or are you testing an experimental layout?

Either way, the current behavior is invalid in this case. There is currently no workaround via options and attributes (except for using another type of phase layout).

**Comment by namdre** [comment]
to fix this this we probably need a new network attribute to distinguish between streams that have priority due to traffic light rules (left vs oncoming) and those that have priority due to the underlying priority junction (left vs cross traffic). The former can than receive special treatment in the stuck-on-junction-after-phase-ends scenario.

**Comment by marecabo** [comment]
Thank you for your detailed response! I see it making sense when the following happens at the same time,

- the east-west-flow turns red
- and the north-south flow turns green

without enough common red time in both flows to allow clearance of the waiting left-turning vehicles in that common red time.

However, I am modelling the [DLR research junction](https://www.dlr.de/ts/en/desktopdefault.aspx/tabid-11367/19949_read-46553) ([OSM](https://www.openstreetmap.org/#map=19/52.27553/10.53562)) with the real traffic light program. In that program, there is a 

- 5 s red for all north-*-flows after
- the east-west-flow switched from orange to red

which is enough time in reality to let the (three at most) vehicles standing on the intersection area to clear it.

**Comment by namdre** [comment]
Let me rephrase my question: Does the following sequence show up in any real-world plan that you are studying?
1) left turning traffic together with oncoming traffic
2) left turn traffic switches to red but oncoming traffic continues

Regarding your observation: Yes. Vehicles waiting on an intersection are eager to leave the intersection. They have no knowledge about the duration of the clearance time (which I suppose is not unrealistic).

**Comment by marecabo** [comment]
> Let me rephrase my question: Does the following sequence show up in any real-world plan that you are studying?
> 
> 1. left turning traffic together with oncoming traffic
> 2. left turn traffic switches to red but oncoming traffic continues

Yes, this exact sequence is part of the real-world traffic light program of the DLR research junction ([OSM](https://www.openstreetmap.org/#map=19/52.27553/10.53562)) on the north-east corner of the Brunswick ring road.

**Comment by marecabo** [comment]
Is there anything I can support you with resolving this issue? Unfortunately, this is quite the crucial behavior for realistic intersection modelling in SUMO. Thank you for all your efforts so far.

**Comment by namdre** [comment]
New modelling options have come up that may be able to satisfy your requirements: In netedit you can now assign a distinct signal state to the internal junction connection (the place where the left turning vehicles wait within the junction). 
In inspect mode with connections shown (Alt+5), set linkIndex2 of the connection to a valid value. 
This way you can control the times where those vehicles may drive explicitly.
You may want to add another state column in tls mode first.

**Comment by namdre** [comment]
also shows up in internal scenario test RiLSA3

**Comment by namdre** [comment]
To make all collisions disappear, the network in the attached issue.zip must be rebuilt (#4397)

**Comment by behrisch** [comment]
In meso the relevant tests show teleports because the vehicles never find a gap (junction control is enabled). I'd consider this expected behavior but would love to hear a second opinion @namdre

**Comment by namdre** [comment]
@behrisch I'd also consider this expected behavior (an expected limitation of the simplified junction model without internal lanes)
